### PR TITLE
fix(rechallenge): Remove JS files from build pipeline when JS is disabled

### LINKED ISSUE #16361: react challenges: retranspiling causes constant "Something went wrong" flash messages
@BerkeleyTrue related to #16305. This is due to babel trying to retranspile on each edit, even when JS is disabled, creating syntax errors as the camper types (e.g. unclosed tags). The stacking messages are intentional, but they should not be occurring during editing. 

[Per Berkeley](https://github.com/freeCodeCamp/freeCodeCamp/issues/16305#issuecomment-354346061):
> I suggest we focus our efforts in disabling the transpile pipeline when js is disabled and to provide a better error message when babel fails to transpile.
<br>

image

### PR DESCRIPTION
**Pull request by Bouncey**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.org/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.
- [x] Addressed currently open issue (replace XXXXX with an issue no in next line)

Closes #16361

### CODE REVIEW

File: common/app/routes/Challenges/redux/execute-challenge-epic.js [hunk 1]
```diff
@@ -41,17 +39,20 @@ export function updateMainEpic(actions, { getState }, { document }) {
...
         // if isCodeLocked do not run challenges
-        .filter(() => (
-          !codeLockedSelector(getState()) &&
-          showPreviewSelector(getState())
-        ))
+        .filter(() => {
+          const state = getState();
+          return (
+            !codeLockedSelector(state) &&
+            isJSEnabledSelector(state) &&
```
Comment (BerkeleyTrue) [inline common/app/routes/Challenges/redux/execute-challenge-epic.js]: This disables the live preview when editing html

File: common/app/routes/Challenges/redux/execute-challenge-epic.js [hunk 2]
```diff
@@ -41,17 +39,20 @@ export function updateMainEpic(actions, { getState }, { document }) {
       const frameMain = createMainFramer(document, getState, proxyLogger);
       const buildAndFrameMain = actions::ofType(
         types.unlockUntrustedCode,
-        types.modernEditorUpdated,
```
Comment (BerkeleyTrue) [inline common/app/routes/Challenges/redux/execute-challenge-epic.js]: This disables the live preview window for modern challenges

File: common/app/routes/Challenges/redux/execute-challenge-epic.js [hunk 3]
```diff
@@ -95,8 +96,14 @@ export function executeChallengeEpic(actions, { getState }, { document }) {
...
       const buildAndFrameChallenge = actions::ofType(types.executeChallenge)
         .debounce(executeDebounceTimeout)
-        // if isCodeLocked do not run challenges
-        .filter(() => !codeLockedSelector(getState()))
+        // if isCodeLocked or JS is disabled, do not run challenges
+        .filter(() => {
+          const state = getState();
+          return (
+            !codeLockedSelector(state) ||
+            isJSEnabledSelector(state)
```
Comment (BerkeleyTrue) [inline common/app/routes/Challenges/redux/execute-challenge-epic.js]: This only runs on execute which sets isJSEnabled to true, so this conditional would never be false

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

**Comment by BerkeleyTrue** [comment]
The best approach for this issue is going to be changing the build pipeline not changes to redux

The entry would be here https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/common/app/routes/Challenges/utils/build.js#L48

This is where we can use state to selectively enabled/disable parts of the pipeline.

Here https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/common/app/routes/Challenges/rechallenge/transformers.js#L100

Is where we can also inject logic to selectively disable transformers.

**Comment by BerkeleyTrue** [comment]
I'm liking where the changes are going. There is a lot of complexity that may not be necessary. 
This still does not actually remove the js transforms from the pipeline when js is disabled.

File: client/epics/err-epic.js [hunk 4]
```diff
@@ -4,7 +4,7 @@ export default function errorSaga(actions) {
   return actions.filter(({ error }) => !!error)
     .map(({ error }) => error)
     .doOnNext(error => console.error(error))
-    .map(() => makeToast({
-      message: 'Something went wrong, please try again later'
+    .map(error => makeToast({
+      message: error.message || 'Something went wrong, please try again later'
```
Comment (BerkeleyTrue) [inline client/epics/err-epic.js]: We shouldn't be displaying err messages to users.

File: common/utils/polyvinyl.js [hunk 5]
```diff
@@ -5,11 +5,11 @@ import castToObservable from '../app/utils/cast-to-observable.js';
 
 
 // createFileStream(
-//   files: Dictionary[Path, PolyVinyl]
+//   files: List[ PolyVinyl ]
```
Comment (BerkeleyTrue) [inline common/utils/polyvinyl.js]: You don't need to include `List` here. To express an array in rtype its just `[]`. To express a typed array it's `[...Polyvinyl]`.

**Comment by BerkeleyTrue** [comment]
@Bouncey I apologies for being latent on this PR but I'm mulling over an idea in my head that hasn't yet reified.

The abstract of it came from your use of the error to signal to the user that the transform has failed.

In the build pipelines current shape this is necessary for this feature to work, but it got me thinking about how we get messages from the frame-runner out to the app state.

I'm leaning toward a solution that would mimic this behavior. Possibly adding a proxy to tap into the build pipeline that would be used to get messages out. This way we wouldn't need to highjack the error handling thread of the observable and we would add the the api of the pipeline itself making it more flexible and powerful.

Does this make sense? What are your thoughts?

**Comment by Bouncey** [comment]
@BerkeleyTrue Sorry I am only seeing this now, I have seeing 'blond, brunette, red-head' for the past week.

I think that having a way to provide helpful messages to the user would be great. At the minute the `friendly message` is just being dumped to the browser console, which is no good if the user relys on the proxy console.

Would you want it done along with this PR or a separate one?

**Comment by Bouncey** [comment]
I've just had a quick play with this idea, and the only way I can see of doing it is setting another property on the poly as they come through to hold the `friendlyError`. Then if that prop is defined, dispatch an action somehow.

At the minute it would require us to inject dispatch from the execute challenge epic

**Comment by Bouncey** [comment]
@BerkeleyTrue Should we merge this is get Beta out the door and put friendly error messages on the road map?

**Review by BerkeleyTrue — APPROVED** [review]

**Comment by BerkeleyTrue** [comment]
Alright, if you think this is ready we'll merge it in.

**Pull request merged**